Preserve SIL when merging modules

slavapestov opened the pull request:
Fixes <rdar://problem/18913977>.

Now that https://github.com/apple/swift/pull/11909 and previous PRs are in, we can finally look up associated types in imported conformances after Sema has been torn down. Phew!

Next steps in this area:
- Emit default argument generators with shared linkage and serialized bit set in Swift 4 mode
- Finish investigating emitting inlineable functions with shared linkage

slavapestov commented:
So @jrose-apple, the problem is that while we constructed a SIL module with the right sections, nothing was forcing them to be read in. Also, I'm going to disable the SIL optimizer when there's no primary file. Does the general approach here look reasonable?

jrose-apple commented:
You can't just disable the SIL optimizer when there's no primary file; that's WMO mode. Really in merge-modules we don't want to run *any* SIL passes—mandatory, optimizing, whatever. I took a manual stab at it at one point by passing `-disable-sil-perf-optzns` but today that still runs the "Onone passes". @swiftix, @gottesmm, would you be okay with -disable-sil-perf-optzns actually turning off *all* the optimization passes?

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -592,26 +594,33 @@ static bool performCompile(std::unique_ptr<CompilerInstance> &Instance,
-    SM->verify();
+  if (!PrimarySourceFile) {
+    SM->linkAll();
+    SM->dump();
```
Clearly a WIP. :-)

atrick replied:
I'm not sure what's being proposed, but `-disable-sil-perf-optzns` should be analogous to `-disable-llvm-optzns`. It should only disable optional SIL passes and should not affect functionality.

jrose-apple replied:
The Onone optimization passes *are* optional. (I'm not talking about the mandatory passes; this is the specific set of "optimizations run even at Onone".)

gottesmm replied:
@atrick That is what I said below. They should add a different flag for this. @jrose-apple This brings up an interesting question. Right now we do not have that concept in our pipeline. Optimizations that run at -Onone are considered mandatory.

atrick replied:
It's well understood that `-Onone` does not do what it says. The `-disable` flag should also disable the Onone optimizations. It's just that no one has bothered implementing it.

jrose-apple commented on include/swift/Serialization/SerializedSILLoader.h:
```diff
@@ -112,6 +112,10 @@ class SerializedSILLoader {
 
   /// Deserialize all SILFunctions, VTables, and WitnessTables for
   /// a given Module.
+  void getAllForModule(Identifier Mod);
```
This is the same as passing `nullptr` for `PrimaryFile` below. I'm not sure if this is better because it's more explicit, or worse because it's easy to do by accident.

slavapestov replied:
No it's the opposite. If you pass a null primary file below, we call getAll(true) on each iteration. I want to call getAll(false).

slavapestov commented:
Ah. My bad on the primary file check. How do I detect that we have _no_ source files?

jrose-apple commented:
I don't know if that's the right check either. Last time I talked to @gottesmm about this we thought it probably made sense to add more SIL stages, and then be more careful about mixing them.

gottesmm commented on include/swift/SIL/SILModule.h:
```diff
@@ -459,6 +459,8 @@ class SILModule {
   /// Link in all VTables in the module.
   void linkAllVTables();
 
+  void linkAll();
```
Comment?

slavapestov replied:
I'm going to redo this and add comments etc. I just wanted to get feedback on the general approach(and it turns out I forgot WMO exists)

gottesmm reviewed (commented):
@jrose-apple I would rather add a separate flag than the perf flag. The reason why is that sometimes it is useful to just disable the perf passes despite doing at -O build.

gottesmm commented:
@slavapestov SGTM! In terms of general approach, IMO greater granularity is in general better. Not having to recompile when debugging is good.

slavapestov committed:
Driver: Pass the new -sil-merge-partial-modules flag

Also pass flags to disable SIL optimization passes when merging
modules, since that's completely unnecessary.

An evolution test that used to fail with WMO disabled now passes
with this change.

FIxes <rdar://problem/18913977>.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
This is now ready for review -- @graydon, @jrose-apple, @gottesmm.

slavapestov commented:
Looks like swift-xcode-playground-support exposed a real bug here.

slavapestov commented:
Reduced test case:

```
import Foundation

func replace(rgx: NSRegularExpression, in: String, with: String, x: NSRange) -> String {
  return rgx.stringByReplacingMatches(in: `in`, options: [], range: x, withTemplate: with)
}
```

It's crashing while deserializing the generic environment of _T0So19NSRegularExpressionC15MatchingOptionsVs10SetAlgebra10FoundationsAEPxqd__cs8SequenceRd__8Iterator_7ElementQYd__AJRtzlufCTW, which is a protocol conformance witness thunk:

```
1.	While deserializing decl #29 (XREF) in 'x'
2.	Cross-reference to module 'Foundation'
	... NSRegularExpression
	... in an extension in module 'Foundation'
	... MatchingOptions
	... Element
```

Of course the code compiles with -WMO or without -sil-merge-partial-modules so the driver or frontend must not be configuring something that would allow the imported decl to be found.

@jrose-apple Any ideas? Should be a simple oversight on my part.

jrose-apple commented:
It could just be another case where SIL ends up pulling in new declarations from the Clang importer too late in the process. This isn't a new issue, but we mostly avoid it today by not having inlinable code that references Clang declarations. I'm not sure that simply declaring a TypeChecker that lives through the whole merge-modules phase is correct either—SILGen and SILOpt are probably set up to assume the set of external declarations doesn't change.

jrose-apple commented:
That is, it's possible there's a whole army of yaks over this hill.

jrose-apple commented on include/swift/AST/SILOptions.h:
```diff
@@ -56,6 +56,11 @@ class SILOptions {
 
+  /// Controls whether to pull in SIL from partial modules during the
+  /// merge modules step. Could perhaps be merged with the link mode
+  /// above but the interactions between all the flags are tricky.
```
I think it's worth trying to merge them, in the spirit of "not leaving work around to clean up later". The settings are already not orthogonal.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -638,6 +638,14 @@ ToolChain::constructInvocation(const MergeModuleJobAction &job,
+
+  // Disable SIL optimization passes; we've already optimized the code in each
+  // partial mode.
+  Arguments.push_back("-disable-diagnostic-passes");
```
\*sigh\* I wish this was inferred, but meanwhile this seems reasonable.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -638,6 +638,14 @@ ToolChain::constructInvocation(const MergeModuleJobAction &job,
+  // Disable SIL optimization passes; we've already optimized the code in each
+  // partial mode.
+  Arguments.push_back("-disable-diagnostic-passes");
+  Arguments.push_back("-disable-sil-perf-optzns");
```
We should change this to also disable running the Onone passes.

gottesmm replied:
@jrose-apple I think in such a case in our pipeline, we need a better separation conceptually in between the "diagnostic passes" and the "mandatory optimizations". Those should be conceptually two different groups of passes. I would rather have a separate flag. Perhaps:

-disable-sil-mandatory-optzns.

And then a third flag that just disables /all/ swift optzns, i.e.:

-disable-sil-optzns

jrose-apple replied:
I'm talking about SILPassPipelinePlan::getOnonePassPipeline. That's not mandatory optimizations.

jrose-apple replied:
Also, "mandatory optimizations" is an oxymoron, and therefore I wouldn't expect `-disable-sil-optzns` to disable the mandatory passes. Skipping the diagnostic passes and the mandatory passes should just be automatic from seeing the SIL stage.

gottesmm replied:
I talked with @jrose-apple offline. Turns out my memory was wrong and the -Onone pass pipeline is not what I thought it was. The -Onone pass pipeline is a pass pipeline that includes extra optimizations run at -Onone.

jrose-apple reviewed (commented):
I'm glad this is mostly coming out pretty simple, though!

slavapestov commented:
For posterity here's a test case that triggers this problem even without my change.

x.swift:

```
import Foundation

@inline(__always)
public func replace(rgx: NSRegularExpression, in: String, with: String, x: NSRange) -> String {
  return rgx.stringByReplacingMatches(in: `in`, options: [], range: x, withTemplate: with)
}
```

y.swift:

```
import Foundation
import x

public func rreplace(rgx: NSRegularExpression, in: String, with: String, x: NSRange) -> String {
  return replace(rgx: rgx, in: `in`, with: with, x: x)
}
```

Compile x.swift with WMO, and y.swift with -O (doesn't matter if you enable WMO or not).

jrose-apple commented:
Maybe we can throw that into the validation suite.

slavapestov commented:
@DougGregor recently did some work on the ClangImporter. Let's try running the tests again.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
New failure, but in the same code:

```
01:11:46 0  swift                    0x0000000112164ee8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
01:11:46 1  swift                    0x0000000112165596 SignalHandler(int) + 454
01:11:46 2  libsystem_platform.dylib 0x00007fff9f740bba _sigtramp + 26
01:11:46 3  libsystem_platform.dylib 000000000000000000 _sigtramp + 1619784800
01:11:46 4  swift                    0x000000010fe3f848 swift::LookUpConformanceInSubstitutionMap::operator()(swift::CanType, swift::Type, swift::ProtocolType*) const + 40
01:11:46 5  swift                    0x000000010fe47b4e llvm::Optional<swift::ProtocolConformanceRef> llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolType*)>::callback_fn<swift::LookUpConformanceInSubstitutionMap>(long, swift::CanType, swift::Type, swift::ProtocolType*) + 14
01:11:46 6  swift                    0x000000010fe3fe28 getMemberForBaseType(llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolType*)>, swift::Type, swift::Type, swift::AssociatedTypeDecl*, swift::Identifier, swift::OptionSet<swift::SubstFlags, unsigned int>) + 648
01:11:46 7  swift                    0x000000010fe43e57 llvm::Optional<swift::Type> llvm::function_ref<llvm::Optional<swift::Type> (swift::TypeBase*)>::callback_fn<substType(swift::Type, llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolType*)>, swift::OptionSet<swift::SubstFlags, unsigned int>)::$_17>(long, swift::TypeBase*) + 1271
01:11:46 8  swift                    0x000000010fe40d83 swift::Type::transformRec(llvm::function_ref<llvm::Optional<swift::Type> (swift::TypeBase*)>) const + 163
01:11:46 9  swift                    0x000000010fe41003 swift::Type::transformRec(llvm::function_ref<llvm::Optional<swift::Type> (swift::TypeBase*)>) const + 803
01:11:46 10 swift                    0x000000010fe3f699 swift::Type::subst(llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolType*)>, swift::OptionSet<swift::SubstFlags, unsigned int>) const + 105
01:11:46 11 swift                    0x000000010f937c5a (anonymous namespace)::SILTypeSubstituter::visitType(swift::CanType) + 90
01:11:46 12 swift                    0x000000010f93774e swift::CanType swift::CanTypeVisitor<(anonymous namespace)::SILTypeSubstituter, swift::CanType>::visit<>(swift::CanType) + 94
01:11:46 13 swift                    0x000000010f93269f (anonymous namespace)::SILTypeSubstituter::substSILFunctionType(swift::CanTypeWrapper<swift::SILFunctionType>) + 623
01:11:46 14 swift                    0x000000010f9323e7 swift::SILFunctionType::substGenericArgs(swift::SILModule&, llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolType*)>) + 71
01:11:46 15 swift                    0x000000010f93231a swift::SILFunctionType::substGenericArgs(swift::SILModule&, swift::SubstitutionMap const&) + 58
01:11:46 16 swift                    0x000000010f93220a swift::SILFunctionType::substGenericArgs(swift::SILModule&, llvm::ArrayRef<swift::Substitution>) + 58
01:11:46 17 swift                    0x000000010f9c18ff (anonymous namespace)::SILVerifier::checkApplySubstitutions(llvm::ArrayRef<swift::Substitution>, swift::SILType) + 319
01:11:46 18 swift                    0x000000010f9c0a3c (anonymous namespace)::SILVerifier::checkFullApplySite(swift::FullApplySite) + 76
01:11:46 19 swift                    0x000000010f9a2277 swift::SILVisitor<(anonymous namespace)::SILVerifier, void>::visit(swift::ValueBase*) + 13735
01:11:46 20 swift                    0x000000010f99d883 (anonymous namespace)::SILVerifier::visitSILBasicBlock(swift::SILBasicBlock*) + 1315
01:11:46 21 swift                    0x000000010f997c97 swift::SILFunction::verify(bool) const + 8823
01:11:46 22 swift                    0x000000010f999d45 swift::SILModule::verify() const + 197
01:11:46 23 swift                    0x000000010ee35100 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 13360
01:11:46 24 swift                    0x000000010edef430 main + 3312
01:11:46 25 libdyld.dylib            0x00007fff9f533255 start + 1
```

slavapestov commented:
@DougGregor looks like it got further, though. This is in SIL verification at the end of merge-modules.

DougGregor commented:
You know, that backtrace looks a whole lot like something fixed by https://github.com/apple/swift/pull/8812.

DougGregor commented:
@swift-ci Please test

DougGregor commented:
No such luck. Needs actually debugging

slavapestov commented:
@swift-ci Please test macOS

slavapestov committed:
Add test case for imported conformance usage after Sema tear-down

This test used to fail because we would look up an associated type
in an incomplete conformance after Sema was torn down, causing a
crash.

Now, the ClangImporter is able to correctly complete the conformance.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov committed:
AST: Use getSuperclassForDecl() in one spot

DougGregor commented:
Meh, you're missing a

  REQUIRES: objc_interop

one of the new tests.

slavapestov commented:
Source compat test suite passed. This new version just adds the REQUIRES:.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
I think this is causing a failure on some of our internal bots -- reverting for now.
